grug/testing/src/client.rs

```diff
@@ -0,0 +1,313 @@
+use {
+    crate::{MakeBlockOutcome, suite::TestSuite},
+    async_trait::async_trait,
+    grug_app::{AppError, Db, Indexer, NaiveProposalPreparer, NullIndexer, ProposalPreparer, Vm},
+    grug_db_memory::MemDb,
+    grug_types::{
+        Binary, Block, BlockClient, BlockInfo, BlockOutcome, BorshDeExt, BroadcastClient,
+        BroadcastTxOutcome, Hash256, Query, QueryClient, QueryResponse, SearchTxClient,
+        SearchTxOutcome, Timestamp, Tx, TxOutcome, UnsignedTx,
+    },
+    grug_vm_rust::RustVm,
+    std::{collections::BTreeMap, ops::DerefMut, sync::Arc, thread, time::Duration},
+    tokio::{runtime::Runtime, sync::Mutex},
+};
+
+pub struct MockClient<DB = MemDb, VM = RustVm, PP = NaiveProposalPreparer, ID = NullIndexer>
+where
+    DB: Db,
+    VM: Vm,
+    PP: ProposalPreparer,
+    ID: Indexer,
+{
+    suite: Arc<Mutex<TestSuite<DB, VM, PP, ID>>>,
+    blocks: Arc<Mutex<BTreeMap<u64, (Block, BlockOutcome)>>>,
+    txs: Arc<Mutex<BTreeMap<Hash256, SearchTxOutcome>>>,
+    block_mode: BlockModeCache,
+}
+
+impl<DB, VM, PP, ID> MockClient<DB, VM, PP, ID>
+where
+    DB: Db + Send + Sync + 'static,
+    VM: Vm + Clone + Send + Sync + 'static,
+    PP: ProposalPreparer + Send + Sync + 'static,
+    ID: Indexer + Send + Sync + 'static,
+    AppError: From<DB::Error> + From<VM::Error> + From<PP::Error> + From<ID::Error>,
+{
+    pub fn new(suite: TestSuite<DB, VM, PP, ID>, block_mode: BlockCreation) -> Self {
+        let suite = Arc::new(Mutex::new(suite));
+        let blocks = Arc::new(Mutex::new(BTreeMap::new()));
+        let txs = Arc::new(Mutex::new(BTreeMap::new()));
+
+        let block_mode = match block_mode {
+            BlockCreation::Timed => {
+                let buffer = Arc::new(Mutex::new(vec![]));
+
+                let th_buffer = buffer.clone();
+                let th_suite = suite.clone();
+                let th_blocks = blocks.clone();
+                let th_txs = txs.clone();
+
+                thread::spawn(move || {
+                    let rt = Runtime::new().unwrap();
+
+                    rt.block_on(async move {
```

Comment:
Since this `async` is only for `lock()`, same as other comments. I think it could be removed and simplified, and `next_block` doesn't need to be `async` either. That'd remove the need for the Tokio runtime use here.